nesta_ds_utils/file_ops.py

```diff
@@ -1,53 +1,196 @@
-        os.remove(zip_path)
\ No newline at end of file
+        os.remove(zip_path)
+
+
+def get_bucket_filenames_s3(bucket_name: str, dir_name: str = "") -> List[str]:
+    """Get a list of all files in bucket directory.
+
+    Args:
+        bucket_name (str): S3 bucket name
+        dir_name (str, optional): Directory or sub-directory within an S3 bucket.
+            Defaults to '' (the top level bucket directory).
+
+    Returns:
+        List[str]: List of file names in bucket directory.
+    """
+    s3_resources = boto3.resource("s3")
+    my_bucket = s3_resources.Bucket(bucket_name)
+    return [
+        object_summary.key
+        for object_summary in my_bucket.objects.filter(Prefix=dir_name)
+    ]
+
+
+def _df_to_fileobj(df: pd.DataFrame, save_file_dir: str) -> io.BytesIO:
+    """Convert DataFrame into bytes file object.
+
+    Args:
+        df (pd.DataFrame): Dataframe to convert.
+        save_file_dir (str): Saving file name.
+
+    Returns:
+        io.BytesIO: Bytes file object.
+    """
+    buffer = io.BytesIO()
+    if fnmatch(save_file_dir, "*.csv"):
+        df.to_csv(buffer)
+    elif fnmatch(save_file_dir, "*.json"):
+        df.to_json(buffer)
+    elif fnmatch(save_file_dir, "*.pkl"):
+        df.to_pickle(buffer)
+    elif fnmatch(save_file_dir, "*.xml"):
+        df.to_xml(buffer)
+    else:
+        raise Exception(
+            "Automatic convertion from pd.DataFrame currently supported only for 'csv', 'json', 'pkl' and 'xml'."
+        )
+    buffer.seek(0)
+    return buffer
+
+
+def _fileobj_to_df(fileobj: io.BytesIO, load_file_dir: str) -> pd.DataFrame:
+    """Convert bytes file object into DataFrame.
+
+    Args:
+        buffer (io.BytesIO): Bytes file object.
+        load_file_dir (str): Loading file name.
+
+    Returns:
+        pd.DataFrame: Dataframe converted
+    """
+    if fnmatch(load_file_dir, "*.csv"):
+        df = pd.read_csv(fileobj)
+    elif fnmatch(load_file_dir, "*.json"):
+        df = pd.read_json(fileobj)
+    elif fnmatch(load_file_dir, "*.pkl"):
+        df = pd.read_pickle(fileobj)
+    elif fnmatch(load_file_dir, "*.xml"):
+        df = pd.read_xml(fileobj)
+    else:
+        raise Exception(
+            "Automatic convertion into pd.DataFrame currently supported only for 'csv', 'json', 'pkl' and 'xml'."
+        )
+    return df
+
+
+def upload_obj(
+    obj: Union[io.BytesIO, pd.DataFrame], bucket: str, file_name_to: str, **kwargs
+):
+    """Uploads data from memory to S3 location.
+
+    Args:
+        obj (Union[io.BytesIO, pd.DataFrame]): Data to upload.
+        bucket (str): Bucket's name.
+        file_name_to (str): Path location to save data.
+
+    Raises:
+        Exception: Util supports data only as "io.BytesIO" or "pd.DataFrame".
+    """
+    s3 = boto3.client("s3")
+    if isinstance(obj, pd.DataFrame):
+        obj = _df_to_fileobj(obj, file_name_to)
+    if isinstance(obj, io.BytesIO):
+        s3.upload_fileobj(obj, bucket, file_name_to, kwargs)
```

Comment:
Does this mean that we're expecting anyone uploading an object that's not a dataframe to have converted it to `BytesIO` first? Seems like that would lead to a lot of repetition which could be handled here?

Also, fwiw, I'm not 100% sure this will work for numpy arrays. I had a quick read the other day and I think they need handling as a special case.

Reply:
My understanding was that `upload_fileobj` of `boto3` only accept `BytesIO` objects, am I wrong? If that's the case then I was thinking in general to expect users to provide `BytesIO`, but having the util also to handle `DataFrame` for convenience.

Of course this means if you want to upload a `DataFrame` you have the choice of either convert it to `BytesIO` yourself, or provide it as `DataFrame`. Is this the kind of repetition you are referring to? 

For numpy arrays, I think you can just `pickle.dump` them into a `BytesIO` buffer for example, am I missing something?